pkg/controller/daemon/update.go

```diff
@@ -64,20 +73,311 @@ func (dsc *DaemonSetsController) rollingUpdate(ds *extensions.DaemonSet) error {
+	var histories []*apps.ControllerRevision
+	var currentHistories []*apps.ControllerRevision
+	histories, err = dsc.controlledHistories(ds)
+	if err != nil {
+		return nil, nil, err
+	}
+	for _, history := range histories {
+		// Add the unique label if it's not already added to the history
+		// We use history name instead of computing hash, so that we don't need to worry about hash collision
+		if _, ok := history.Labels[extensions.DefaultDaemonSetUniqueLabelKey]; !ok {
+			var clone interface{}
+			clone, err = api.Scheme.DeepCopy(history)
+			if err != nil {
+				return nil, nil, err
+			}
+			toUpdate := clone.(*apps.ControllerRevision)
+			toUpdate.Labels[extensions.DefaultDaemonSetUniqueLabelKey] = toUpdate.Name
+			history, err = dsc.kubeClient.AppsV1beta1().ControllerRevisions(ds.Namespace).Update(toUpdate)
+			if err != nil {
+				return nil, nil, err
+			}
+		}
+		// Compare histories with ds to separate cur and old history
+		found := false
+		found, err = Match(&ds.Spec.Template, history)
+		if err != nil {
+			return nil, nil, err
+		}
+		if found {
+			currentHistories = append(currentHistories, history)
+		} else {
+			old = append(old, history)
+		}
+	}
+
+	currRevision := maxRevision(old) + 1
+	switch len(currentHistories) {
+	case 0:
+		// Create a new history if the current one isn't found
+		cur, err = dsc.snapshot(ds, currRevision)
+		if err != nil {
+			return nil, nil, err
+		}
+	default:
+		cur, err = dsc.dedupCurHistories(ds, currentHistories)
+		if err != nil {
+			return nil, nil, err
+		}
+		// Update revision number if necessary
+		if cur.Revision < currRevision {
+			var clone interface{}
+			clone, err = api.Scheme.DeepCopy(cur)
+			if err != nil {
+				return nil, nil, err
+			}
+			toUpdate := clone.(*apps.ControllerRevision)
+			toUpdate.Revision = currRevision
+			_, err = dsc.kubeClient.AppsV1beta1().ControllerRevisions(ds.Namespace).Update(toUpdate)
+			if err != nil {
+				return nil, nil, err
+			}
+		}
+	}
+	// Label ds with current history's unique label as well
+	if ds.Labels[extensions.DefaultDaemonSetUniqueLabelKey] != cur.Labels[extensions.DefaultDaemonSetUniqueLabelKey] {
+		var clone interface{}
+		clone, err = api.Scheme.DeepCopy(ds)
+		if err != nil {
+			return nil, nil, err
+		}
+		toUpdate := clone.(*extensions.DaemonSet)
+		if toUpdate.Labels == nil {
+			toUpdate.Labels = make(map[string]string)
+		}
+		toUpdate.Labels[extensions.DefaultDaemonSetUniqueLabelKey] = cur.Labels[extensions.DefaultDaemonSetUniqueLabelKey]
+		_, err = dsc.kubeClient.ExtensionsV1beta1().DaemonSets(ds.Namespace).UpdateStatus(toUpdate)
+	}
+	return cur, old, err
+}
+
+func (dsc *DaemonSetsController) cleanupHistory(ds *extensions.DaemonSet) error {
+	nodesToDaemonPods, err := dsc.getNodesToDaemonPods(ds)
+	if err != nil {
+		return fmt.Errorf("couldn't get node to daemon pod mapping for daemon set %q: %v", ds.Name, err)
+	}
+	_, old, err := dsc.constructHistory(ds)
```

Comment:
I am wondering if it's going to be cleaner to construct the history once and then pass the references in `manage` and `cleanupHistory`, instead of reconstructing it for each one of them.